src/options.ts

```diff
@@ -27,58 +27,67 @@ Plot.valueof is not guaranteed to return a new array. When a transform method is
+export function valueof<N extends null | undefined>(data: N, value: ValueAccessor<N>): N;
+export function valueof<N extends Datum>(
+  data: Data<N> | null | undefined,
+  value: ValueAccessor<N>,
   type: Float64ArrayConstructor
 ): Float64Array;
-export function valueof(
-  data: Data | null | undefined,
-  value: ValueAccessor,
+export function valueof<N extends Datum>(
+  data: Data<N> | null | undefined,
+  value: ValueAccessor<N>,
   type: Float32ArrayConstructor
 ): Float32Array;
-export function valueof(data: Data | null | undefined, value: ValueAccessor, type: ArrayConstructor): ValueArray;
-export function valueof(
-  data: Data | null | undefined,
-  value: ValueAccessor,
+export function valueof<N extends Datum>(
+  data: Data<N> | null | undefined,
+  value: ValueAccessor<N>,
+  type: ArrayConstructor
+): ValueArray;
+export function valueof<N extends Datum>(
+  data: Data<N> | null | undefined,
+  value: ValueAccessor<N>,
   arrayType?: ArrayType
 ): ValueArray | Float32Array | Float64Array | null | undefined;
-export function valueof(
-  data: Data | null | undefined,
-  value: ValueAccessor,
+export function valueof<N extends Datum>(
+  data: Data<N> | null | undefined,
+  value: ValueAccessor<N>,
   arrayType?: ArrayType
 ): ValueArray | Float32Array | Float64Array | null | undefined {
-  return data == null
-    ? data
-    : typeof value === "string"
-    ? map(data, field(value), arrayType)
-    : typeof value === "function"
-    ? map(data, value, arrayType)
-    : typeof value === "number" || value instanceof Date || typeof value === "boolean"
-    ? map(data, constant(value), arrayType)
-    : value && typeof (value as TransformMethod).transform === "function"
-    ? arrayify((value as TransformMethod).transform(data), arrayType)
-    : arrayify(value as ValueArray, arrayType); // preserve undefined type
+  if (data == null) {
+    return data;
+  } else if (typeof value === "string") {
+    return map(data, field(value), arrayType);
+  } else if (typeof value === "function") {
+    return map(data, value, arrayType);
+  } else if (typeof value === "number" || value instanceof Date || typeof value === "boolean") {
+    return map(data, constant(value), arrayType);
+  } else if (isTransform(value)) {
+    return arrayify(value.transform(data), arrayType); // Not sure what's wrong here
+  }
+  return arrayify(value, arrayType); // preserve undefined type
+}
+
+function isTransform<T extends Datum>(value: ValueAccessor<T>): value is TransformMethod<T> {
+  return !!value && typeof value == "object" && "function" in value;
 }
 
 /**
  * See Plot.valueof()
  */
-export type ValueAccessor =
-  | string
-  | AccessorFunction
+export type ValueAccessor<T extends Datum> =
+  | DatumKeys<T>
+  | AccessorFunction<T>
   | number
   | Date
   | boolean
-  | TransformMethod
-  | ValueArray
+  | TransformMethod<T>
+  // | ValueArray // This didn't seem right, how would we use ValueArray as an accessor?
```

Comment:
It's when you pass a "column" as the value accessor.

Typically: `Plot.text({length:1}, {x: [1], y: [2], text: ["Hello"]})`